[Merged by Bors] - Add automatic docs deployment to GitHub Pages

james7132 opened the pull request:
# Objective

Partially address #407 by setting up automated deployments of `bevy`'s API reference to GitHub Pages.

## Solution

Set up a GitHub Actions workflow that builds the docs on every push to `main` and pushes a new commit to a `gh-pages` (or `docs` branch).

A few smaller additions to better address #407:

 - A top level redirect was added to take "docs.bevyengine.org" directly to the `bevy` crate docs.
 - A GitHub Pages CNAME file for supporting a publicly viewable domain instead of `github.io`
 - A robots.txt file is added to disable all search engine crawlers that respect it from crawling it (avoid having conflicting Google search results)
 - A .nojekyl file to speed up deployments since there is no Jekyll templating in the output.

This may require configuration of the  `GITHUB_TOKEN` of the CI to successfully run this.

## Followup

For this to completely resolve #407, a subdomain of https://bevyengine.org/ needs to be set up to point to the CNAME location. This is initially set to "dev-docs.bevyengine.org".

alice-i-cecile reviewed (approved):
This will be very useful. The workflow looks reasonable to me, and it's very clear.

Nilirad commented:
Great! This will close bevyengine/bevy-website#181.

mockersf reviewed (changes requested):
I think it needs to be made very clear in the docs deployed that this would be the doc for Bevy main and not the released version.
Maybe a custom logo with "main" added? Or a sticky header bar that would say "Bevy main - version under development"?

Nilirad commented:
A message on top could say "Development build docs. Click for the latest release docs", or something like that.

Alternatively, we can have a dropdown menu like docs.rs to choose between development and release docs.

james7132 commented:
Added a sed command that changes the icon out for something more noticeable., and added a header.html shim to force Google not to index, even when linked from other sites. Other search engines should basically stop at the robots.txt.

NiklasEi commented:
Before more people go looking: the new icon is coming in bevyengine/bevy-website#234

james7132 commented:
With https://github.com/bevyengine/bevy-website/pull/234 merged, @mockersf, any other thing that's needed? It's a little hard to notice right now, but it should be enough to distinguish between it and docs.rs. I could add a sticky banner as suggested earlier, but `--html-before-content` seems to shove it in a row-oriented flexbox due to the navbar. Not familiar enough with HTML/CSS to make it actually stick above everything.

mockersf commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,53 @@
+        run: cargo doc --all-features --no-deps -p bevy
+
+      #  This adds the following:
+      #   - A top level redirect to the bevy crate docmentation
```
```suggestion
      #   - A top level redirect to the bevy crate documentation
```

mockersf commented:
This should be good to go now.

The target domain/url is something that is up to Cart. Maybe it should not live on the website / under a `bevy engine.org` url as we try to keep the website only for the released version? Maybe a subdomain would be enough to differentiate?

I think the url where it would be deployed without custom domain would be https://bevyengine.github.io/bevy, that could help with having website for release, GitHub for in development?

cart commented:
Yeah I'm ready to move forward on this. I do think that `docs.bevyengine.org` is the wrong name for "in development / unstable" docs. I do think a github url is a reasonable way to correlate the context to "in development", but its worth calling out that github also hosts our "stable releases and branches". We might be able to make it even more clear with a custom url. `unstable-docs.bevyengine.org`, `dev-docs.bevyengine.org`, `dev-api.bevyengine.org` etc.

I think I like the `dev-docs` prefix, for symmetry with docs.rs. I am interested to hear what everyone else thinks.

james7132 commented:
I agree with the `dev-docs` subdomain. It seems recognizable at first glance, shorter than `unstable-docs`, and doesn't look like a REST API subdomain, unlike `dev-api`.

cart commented:
Cool lets roll with that. I'll get the domain set up.

cart commented:
We might actually need to merge this pr first. Can you update the domain?

cart commented:
bors r+
